packages/base-zone/src/prepare-attenuator.js

```diff
@@ -76,56 +112,59 @@ export const prepareAttenuatorMaker = (
 ) => {
   const {
     uInterfaceName = uKindName,
-    extraMethodGuards = undefined,
-    extraMethods = undefined,
+    extraMethodGuards = {},
+    extraMethods = {},
   } = options;
   const AttenuatorI = M.interface(
     `${uInterfaceName}_attenuator`,
-    {
-      ...extraMethodGuards,
-    },
-    {
-      defaultGuards: 'raw',
-    },
+    extraMethodGuards,
+    { defaultGuards: 'raw' },
   );
 
   const attenuatorKindName = `${uKindName}_attenuator`;
 
   return zone.exoClass(
     attenuatorKindName,
     AttenuatorI,
-    underlying => ({
-      underlying,
-    }),
+    underlying => ({ underlying }),
+    wrapperMethods(attenuatorKindName, uMethodNames, extraMethods),
     {
-      ...fromUniqueEntries(
-        uMethodNames.map(name => [
-          name,
-          {
-            // Use concise method syntax for exo methods
-            [name](...args) {
-              // @ts-expect-error normal exo-this typing confusion
-              const { underlying } = this.state;
-
-              // Because attenuators still support someone adding `selfRevoke`
-              // as an `extraMethod`, this test is still useful. See the
-              // testcase in `prepare-attenuator.test.js`.
-              underlying !== undefined ||
-                Fail`${q(attenuatorKindName)} revoked`;
-
-              return underlying[name](...args);
-            },
-            // @ts-expect-error using possible symbol as index type
-          }[name],
-        ]),
-      ),
-      ...extraMethods,
-    },
-    {
-      stateShape: {
-        underlying: M.opt(M.remotable('underlying')),
-      },
+      stateShape: { underlying: M.opt(M.remotable('underlying')) },
     },
   );
 };
 harden(prepareAttenuatorMaker);
+
+const PrefixedKindNameRE = /(alleged: |DebugName: )(.*)/;
+
+/**
+ * A convenience above `prepareAttenuatorMaker` for doing a singleton
+ * ephemeral attenuator, i.e., a new attenuator instance (and hidden class)
+ * that is allocated in the heap zone, and therefore ephemeral. The underlying
+ * can be durable or not, with no conflict with the attenuator being
+ * ephemeral. The price of this convenience is the allocation of the hidden
+ * class and wrapper methods per `attenuateOne` call, i.e., per attenuator
+ * instance.
+ *
+ * @template {any} U
+ * @param {U} underlying
+ * @param {PropertyKey[]} uMethodNames
+ * @param {AttenuatorOptions<U>} [options]
+ * @returns {Partial<U>}
```

Comment:
Would it be worthwhile to narrow this?
```suggestion
 * @template {Record<string|symbol, (...args: unknown[]) => unknown>} U
 * @template {readonly (keyof U)[]} Picks
 * @template {unknown | Record<string|symbol, (...args: unknown[]) => unknown>} [ExtraMethods]
 * @param {U} underlying
 * @param {Picks} uMethodNames
 * @param {AttenuatorOptions & {extraMethods?: ExtraMethods}} [options]
 * @returns {Pick<Attenuated<U, PermitFromKeys<U, Picks>>, Picks[number]> & { [K in keyof ExtraMethods]: ExtraMethods[K] }}
```
with Attenuated from [packages/internal/src/types.ts](https://github.com/Agoric/agoric-sdk/blob/1f93cea5771408c025ede9abbba6138bf913db29/packages/internal/src/types.ts#L18-L37), and adding
```js
/**
 * @template T
 * @template {readonly (keyof T)[]} Keys
 * @typedef {{ [K in keyof T]: K extends string ? K extends Keys[number] ? true : never : never }} PermitFromKeys
 */
```

Reply:
That whole `Permit` thing is about deeply picking, which is exactly what I'm not doing in this PR. Instead I just changed `PropertyKey[]` here to `(keyof U)[]` to match the previous change.

Reply:
The kind of rich type refinements you have in mind is outside my expertise, and in any case can happen after this PR.

Reply:
Borrowing from your code, I tries introducing
```js
 * @typedef {Record<string, (...args: Passable[]) => Passable>} Methods
```

and then using it everywhere in this module where I currently say
```js
* @param {any} U
```

but it gave me errors I did not understand, so I backed out of that as well.

Reply:
Resolving